README.md
=========================
README for RIRI5
=========================

This is RIRI5 model, an adaptation of the RIRI model in python for simple sky discretisations (diffuse sky) and multi-species canpies

See 
Sinoquet, H., Le Roux, X., Adam, B., Ameglio, T., & Daudet, F. A. (2001). RATP: a model for simulating the spatial distribution of radiation absorption, transpiration and photosynthesis within canopies: application to an isolated tree crown. Plant, Cell & Environment, 24(4), 395-406.
Louarn, G., Escobar-Gutiérrez, A., Migault, V., Faverjon, L., & Combes, D. (2014). “Virtual grassland”: an individual-based model to deal with grassland community dynamics under fluctuating water and nitrogen availability. The future of european grasslands, 242.



## 1. Getting Started

These instructions will get you a copy of *L-egume* up and running on your local 
machine.

### 1.1 Prerequisites

To install and use *RIRI5*, you need first to install the dependencies.

*RIRI5* has been tested on Windows.
 
#### 1.1.1 Install the dependencies on Windows 10 64 bit

1. Install Python  

    * go to https://www.python.org/downloads/windows/download, 
    * click on "Latest Python 2 Release [...]", 
    * download "Windows x86-64 MSI installer" and install it selecting the following options:
        * install for all users,
        * default destination directory,
        * install all subfeatures, including subfeature "Add python.exe to Path".

2. Install OpenAlea Vplants:  

    * go to http://openalea.gforge.inria.fr/dokuwiki/doku.php?id=download:windows, 
    * download `OpenAlea 1.2` and `Vplants 1.2`,
    * install both installers: 
		  
3. Install NumPy:  

    * go to http://www.lfd.uci.edu/~gohlke/pythonlibs/#numpy, 
    * download `NumPy+MKL` for Python 2 64 bit,
    * install it using `pip` installer: 
        * open a command line interpreter,
        * go to the directory where you saved `NumPy+MKL` for Python 2 64 bit,
        * install `NumPy+MKL` from the downloaded wheel file.  
          For example, if you downloaded file "numpy-1.13.1+mkl-cp27-cp27m-win_amd64.whl", 
          type: `pip install "numpy-1.13.1+mkl-cp27-cp27m-win_amd64.whl"`.

4. Install Pandas  

    * go to http://www.lfd.uci.edu/~gohlke/pythonlibs/#pandas, 
    * download `Pandas` for Python 2 64 bit,
    * install it using `pip` installer: 
        * open a command line interpreter,
        * go to the directory where you saved `Pandas` for Python 2 64 bit,
        * install `Pandas` from the downloaded wheel file.  
          For example, if you downloaded file "pandas-0.20.3-cp27-cp27m-win_amd64.whl", 
          type: `pip install "pandas-0.20.3-cp27-cp27m-win_amd64.whl"`.
          
5. Install Sphinx

    * go to http://www.lfd.uci.edu/~gohlke/pythonlibs/#misc, 
    * download `Sphinx` for Python 2,
    * install it using `pip` installer: 
        * open a command line interpreter,
        * go to the directory where you saved `Sphinx` for Python 2,
        * install `Sphinx` from the downloaded wheel file.  
          For example, if you downloaded file "Sphinx-1.6.3-py2.py3-none-any.whl", 
          type: `pip install "Sphinx-1.6.3-py2.py3-none-any.whl"`.
          
### 1.2 Installing

__Note__: We suppose you already installed the dependencies for your operating system. Otherwise follow these [instructions](prerequisites "Prerequisites").

You can install *L-grass* either in "install" or "develop" mode.

#### 1.2.1 Install *L-grass* in "install" mode

Install *L-grass* in "install" mode if you're not going to develop, edit or debug 
it, i.e. you just want to used it as third party package.

To install *RIRI5* in "end-user" mode:

* open a command line interpreter,
* go to your local copy of project *RIRI5*,
* run command: `python setup.py install --user`.

#### 1.2.2 Install *RIRI5* in "develop" mode

Install *RIRI5* in "develop" mode if you want to get *RIRI5* installed and then 
be able to frequently edit the code and not have to re-install *RIRI5* to have the 
changes to take effect immediately.

To install *RIRI5* in "develop" mode:

* open a command line interpreter,
* go to your local copy of project *RIRI5*,
* run command: `python setup.py develop --user`.

### 1.3 Running

To run a simulation example, two options:

* 1. open Lpy platform,
	 load lgrass.lpy file from lgrass folder,
	 Use Run or animate button to launch a simulation
  2. Run lgrass from a python script (see main.py in test folder for an example)

See the user guide for a step by step explanation of how to set and run model *L-egume*.

## 2. Reading the docs

To build the user and reference guides:

* install the model (see [Installation of the model](installing "Installing")), 
* open a command line interpreter,
* go to the top directory of your local copy of the project,
* run this command: `python setup.py build_sphinx`,
* and direct your browser to file `doc/_build/html/index.html`.

## 3. Testing

The test allows to verify that the model implementation accurately 
represents the developer’s conceptual description of the model and its solution.

The test:

* runs the model on 200 steps,
* concatenates the outputs of the model in pandas dataframes,
* writes the outputs dataframes to CSV files,
* compares actual to expected outputs,
* raises an error if actual and expected outputs are not equal up to a given tolerance.     

To run the test :

* install the model (see [Installation of the model](installing "Installing")), 
* open a command line interpreter,
* go to the directory `test` of your local copy of the project,
* and run this command: `python test_lgrass.py`.

## Built With

* [Python](http://www.python.org/), [NumPy](http://www.numpy.org/), [SciPy](http://www.scipy.org/), 
* [Sphinx](http://sphinx-doc.org/): building of the documentation, 

## Contributing

First, send an email to <lgrass-request@groupes.renater.fr> to be added to the project.  

Then,
 
* check for open issues or open a fresh issue to start a discussion around a
  feature idea or a bug: https://sourcesup.renater.fr/tracker/?group_id=3957.
* If you feel uncomfortable or uncertain about an issue or your changes, feel
  free to email <lgrass@groupes.renater.fr>.

## Contact

For any question, send an email to <lgrass-request@groupes.renater.fr>.

## Versioning

We use a Git repository on [SourceSup](https://sourcesup.renater.fr) for 
versioning: https://sourcesup.renater.fr/projects/riri5/.  
If you need an access to the current development version of the model, please send 
an email to <lgrass-request@groupes.renater.fr>.
For versionning, use a git client and get git clone git+ssh://git@git.renater.fr:2222/lgrass.git. SSH will is required

## Authors

**Gaëtan LOUARN**, **Didier Combes** - see file [AUTHORS](AUTHORS) for details

## License

This project is licensed under the CeCILL-C License - see file [LICENSE](LICENSE) for details


CHANGES
RIRI5 1.1
--------------------
- compatible python 3 (3.7)

RIRI5 1.0
--------------------

First version of RIRI5.


riri5/test/test_riri.py
from scipy import *
import riri5 as riri

import RIRI5 as riri #celui du dossier l-egume : qui marche


#test 1 :
m = array([[0.,1.,3.,3.,1.,0.],[0.,1.,3.,3.,1.,0.], [0.,1.,3.,3.,1.,0.], [0.,1.,3.,3.,1.,0.]])
ma = array([m, m,m, m])/4.
ma2 = array([m, m,m, m])/8.
ls_mlai = array([ma, ma2])


ves = array([ma[:,2,2], ma[:,1,1]]) #liste de lai par especeet voxel sur le chemin du rayon
ks = array([1., 0.4]) #liste de k par espece
I0 = 1000.

#pour 1 rayon
res_trans, res_abs_i = riri.calc_extinc_ray_multi(ves, I0, ks) # proche d'un lai de 4 : 1000.*exp(-1.*4.)


#calcul des listes de triplets selon la matrice
triplets = riri.get_ls_triplets(ma, opt='V')
triplets = riri.get_ls_triplets(ma, opt='VXpXmYpYm')


#distributtions d'angles
ls_distf = [riri.disttetaf(abs(45.), 0.), riri.disttetaf(abs(45.), 0.)]

#pour tous les rayons
res_trans_form, res_abs_i_form = riri.calc_extinc_allray_multi(ls_mlai, triplets ,ls_distf , I0)




#test 2: calcul sur matrice creuse avec toutes les cases
ma3 = zeros([8,4,4])
ma3[6:8,:,:] = 1.
ma4 = zeros([8,4,4])
ma4[5:8,:,:] = 0.5
ls_mlai2 = array([ma3, ma4])


triplets = riri.get_ls_triplets(ma3, opt='VXpXmYpYm')
ls_distf = [riri.disttetaf(abs(45.), 0.), riri.disttetaf(abs(45.), 0.)]
res_trans_form, res_abs_i_form = riri.calc_extinc_allray_multi(ls_mlai2, triplets ,ls_distf , I0)


#test 3: avec fonction qui redimensionne (plus rapide!)

res_trans_form, res_abs_i_form = riri.calc_extinc_allray_multi_reduced(ls_mlai2, triplets ,ls_distf , I0, optsky=None, opt='VXpXmYpYm')


riri5/RIRI5.py
##################
##
## Turbid medium 3D - simplifie pour turtle 5 directions (diffus uniquement) - adapted from RATP model
## G. Louarn / Mai 2014
##
##################



from numpy import linspace
from scipy import array, exp, zeros, ones, set_printoptions, pi, radians, cos, sin, tan, arccos, histogram
from copy import deepcopy
from numpy.random import seed, normal
import numpy as np

def get_lsparami(ParamP, param):
    """ recupere une liste des parametre param de chaque plante de L-egume """
    v = []
    for i in range(len(ParamP)):
        v.append(ParamP[i][param])
    return v
# a caser sans doute dans un autre fichier

def def_na_lims(pattern8, dz_ini, Hmax,opt='3D'):
    """ calcul nb de voxels par direction et limites des voxels dans la grille #pattern, dz et Hmax en cm"""
    if opt=='1D':
        nxa,nya = 1,1
        nza = int(Hmax/dz_ini)+1
        dxa,dya = pattern8[1][0] - pattern8[0][0], pattern8[1][1] - pattern8[0][1]
        dza = dz_ini
    elif opt=='3D':
        nxa = int((pattern8[1][0] - pattern8[0][0])/(2*dz_ini))
        nya = int((pattern8[1][1] - pattern8[0][1])/(2*dz_ini))
        dxa = (pattern8[1][0] - pattern8[0][0])/float(nxa)
        dya = (pattern8[1][1] - pattern8[0][1])/float(nya)
        dza = 0.5*dxa #dz recalcule pour 
        nza = int(Hmax/dza)+1

    xlims = linspace(pattern8[0][0], pattern8[1][0],nxa+1)
    ylims = linspace(pattern8[0][1], pattern8[1][1],nya+1)
    zlims = array([x * dza for x in range(0, nza+1)])

    na = [nxa,nya,nza]
    lims = [xlims,ylims,zlims]
    dxyz = [dxa,dya,dza]
    origin_grid = array([pattern8[0][0], pattern8[0][1],zlims[-1]])
    surf_refVOX = dxa*dya/10000.#m2
    return na, dxyz, lims, origin_grid, surf_refVOX

#na, dxyz, lims, origin_grid = def_na_lims(pattern8, dz_aerien, Hmax)


## dans quel voxel?
def WhichVoxel(p, origin_grid, na, dxyz):
    """ """
    """ en z, id=0 = haut du couvert """
    p1_rel = p - origin_grid
    vox = [int(p1_rel[0]//dxyz[0]), int(p1_rel[1]//dxyz[1]), int(-p1_rel[2]//dxyz[2])] #// division entiere

    #test si dans grille et sinon retouve indice correspondant
    test = [0 <= vox[0] <na[0], 0 <= vox[1] <na[1], 0 <= vox[2] <na[2]]
    if test != [True,True,True]:#si hors grille
        #recup les indices du fautif
        matches = [i for i in range(3) if test[i]==False]
        #faire le calcul du bon id equivalent
        for i in matches:
            if i!=2: #x,y -> couvert infini
                vox[i] = vox[i]%na[i] # marche pour > et <0 et na[i]=1
            else: #z ->  
                if vox[i]<0:#au dessus: met dans la derniere strate
                    vox[i] = 0
                else: #en dessous: met dans la strate au dessus du sol
                    vox[i] = na[i]-1
    return vox


#na
#WhichVoxel(array([-12.,-10.,130.]), origin_grid, na, dxyz)
#WhichVoxel(array([-200.,100.,230.]), origin_grid, na, dxyz)#ok
#WhichVoxel(array([-200.,10.,10.]), origin_grid, na, dxyz)#ok
#WhichVoxel(array([-200.,10.,208.]), origin_grid, na, dxyz)#ok



def get_tripletY(ydeb, x, ny, nz, sens='+'):
    """ pour un plan dans y, rayon allant dans le + ou le - """
    zz = list(range(nz))
    xx = [x]*nz
    res = []
    count = ydeb
    if sens=='+':
        for k in range(nz):
            if count > ny-1:
                res.append(0)
                count=1
            else:
                res.append(count)
                count = count+1

    elif sens=='-':
        for k in range(nz):
            if count < 0:
                res.append(ny-1)
                count=ny-2
            else:
                res.append(count)
                count=count-1

    return [zz, res, xx]

#get_tripletY(3, 0, 4, 4, '-')
#get_tripletY(3, 0, 4, 4, '+')
#get_tripletY(0, 0, 4, 4, '-')
#get_tripletY(0, 0, 4, 4, '+')
#get_tripletYbis(0, 0, 4, 10, sens='+')

#utilise un count (initialiser avec y deb) et meme demarche
def get_tripletX(xdeb, y, nx, nz, sens='+'):
    """ pour un plan dans y, rayon allant dans le + ou le - """
    zz = list(range(nz))
    yy = [y]*nz
    res = []
    count = xdeb
    if sens=='+':
        for k in range(nz):
            if count > nx-1:
                res.append(0)
                count=1
            else:
                res.append(count)
                count=count+1
    elif sens=='-':
        for k in range(nz):
            if count < 0:
                res.append(nx-1)
                count=nx-2
            else:
                res.append(count)
                count=count-1

    return [zz, yy, res]


#get_tripletX(3, 0, 6, 4, '+')
#get_tripletX(1, 0, 6, 4, '+')
#get_tripletX(1, 0, 6, 4, '-')
#get_tripletX(0, 2, 4, 10, sens='-')

def get_tripletVert(x,y,nz):
    """ pour triplets verticaux """
    xx = [x]*nz
    yy = [y]*nz
    zz = list(range(nz))
    return [zz,yy,xx]



def get_ls_triplets(mat, opt='VXpXmYpYm'):
    """ liste de triplets par direction; V= vertical; Xp,Yp,Xm,Ym 4 rayons a 45 degre dans le plan normal aux faces des voxels"""
    nz,ny,nx = mat.shape
    ls_tripletYp, ls_tripletYm =[], []
    ls_tripletXp, ls_tripletXm =[], []
    ls_tripletVert = []
    for i in range(nx):
        for j in range(ny):
            ls_tripletYp.append( get_tripletY(j, i, ny, nz, '+') )
            ls_tripletYm.append( get_tripletY(j, i, ny, nz, '-') )
            ls_tripletXp.append( get_tripletX(i, j, nx, nz, '+') )
            ls_tripletXm.append( get_tripletX(i, j, nx, nz, '-') )
            ls_tripletVert.append(get_tripletVert(i,j,nz))
    
    ls_triplet = []
    if opt.find('V')>=0:
        ls_triplet.append(ls_tripletVert)
    if opt.find('Xp')>=0:
        ls_triplet.append(ls_tripletXp)
    if opt.find('Xm')>=0:
        ls_triplet.append(ls_tripletXm)
    if opt.find('Yp')>=0:
        ls_triplet.append(ls_tripletYp)
    if opt.find('Ym')>=0:
        ls_triplet.append(ls_tripletYm)
    
    return ls_triplet

#triplets = get_ls_triplets(ma, opt='V')
#triplets = get_ls_triplets(ma, opt='VXpXmYpYm')


def calc_extinc_ray_multi(ls_vlai, I0, ls_k):
    """ calcul extinction (profil de Iout) et absorption par espece pour un rayon traversant une serie de voxel en serie decrite par profils de LAI (du haut en bas) et k """
    """ pour partage entre especes dans un voxel homogene  - sinoquet et al. 2000 """
    """ prend un k par entite/espece, mais peut facilement prendre profils de k """

    I = I0
    res_trans = []
    res_abs_i = []
    for strate in range(len(ls_vlai[0])):
        lai_ki = ls_vlai[:,strate] * ls_k
        sum_lai_ki = sum(lai_ki)
        if sum_lai_ki!=0. :
            LIEi = lai_ki/sum_lai_ki * (1 - exp(-sum_lai_ki))
        else: #si tout les lai sont a zero, renvoir zero
            LIEi = lai_ki

        Iout = I*(1-sum(LIEi))

        res_trans.append (Iout)
        res_abs_i.append(I*LIEi)
        I=Iout

    res_abs_i = array(res_abs_i).T #chaque ligne correspond a une entite/espece dans l'odre des ls_vlai
    return res_trans,  res_abs_i



#m = array([[0.,1.,3.,3.,1.,0.],[0.,1.,3.,3.,1.,0.], [0.,1.,3.,3.,1.,0.], [0.,1.,3.,3.,1.,0.]])
#ma = array([m, m,m, m])/4.
#ma2 = array([m, m,m, m])/8.

#ves = array([ma[:,2,2], ma[:,1,1]]) #liste de lai par especeet voxel sur le chemin du rayon
#ks = array([1., 0.4]) #liste de k par espece
#I0 = 1000.
#res_trans, res_abs_i = calc_extinc_ray_multi(ves, I0, ks) # proche d'un lai de 4 : 1000.*exp(-1.*4.)


def calc_extinc_allray_multi(ls_mlai, ls_triplets_dir ,ls_distf , I0, optsky=None):
    """ calcul extinction (profil de Iout) et absorption par espece pour des liste de voxel, regroupe par direction """
    """ presupose que ls_triplets tout ou partie d'un turtle 6 => dx=dy=dz / tan(0.4637) = 2*dz !"""
    #res = deepcopy(ls_triplets_dir) #liste des triplets de voxel par direction, associe a res_trans et res_abs_i

    alfa_turtle6 = 0.4637 #radians / 26.57 degre
    # distribution de I0 entre sources
    n_dir = len(ls_triplets_dir)
    if optsky==None or n_dir<5: #cas par defaut = toutes directions le meme poids / a utiliser notamment qd veut tester une seule dir
        ls_poids = array([1./float(n_dir)]*n_dir)  
    elif n_dir == 5 and optsky=='uoc': #diffus couvert / poids pour un turtle 6 ramene a 5 direction
        ls_poids = array([1./6.]+[5./(6.*4)]*4)
    elif n_dir == 5 and optsky=='soc': #diffus clair
        ls_poids = [1./6.]+[5./(6.*4)]*4
        effet_sin = [sin(pi/2)]+ [sin(alfa_turtle6)]*4
        x = array(ls_poids)*effet_sin
        ls_poids = x/sum(x)

    #print 'ls poids:', ls_poids

    #coeff de ponderation des lai par longueur de parcours dans le voxel
    # -> juste un effet dz (du coup dans mon cas tout dz est parcourru)(effet longueur proprement dit integre dans le k)

    #coeff d'extinction par direction et espece kteta
    ls_k_teta = []
    for i in range(len(ls_distf)):
        ls_k_teta.append([k_teta_distf(90., ls_distf[i]), k_teta_distf(26.57, ls_distf[i])])

    ls_k = []
    for i in range(len(ls_distf)):
        if ls_triplets_dir[0][0][1] == ls_triplets_dir[0][0][2]: #y a 1 ray vertical, place en premier selon get_ls_triplets
            ls_k.append([ls_k_teta[i][0]] + [ls_k_teta[i][1]]*(n_dir-1))
        else:
            ls_k.append([ls_k_teta[i][1]]*n_dir)

    ls_k = array(ls_k)
    #print 'ls k:', ls_k

    #calculs
    res_trans_form = zeros(ls_mlai[0].shape)#sortie pour recuperer les cumuls trans
    res_abs_i_form  = zeros(ls_mlai.shape)#sortie pour recuperer les cumul abs par espece
    for dir in range(len(ls_triplets_dir)):
        for tri in range(len(ls_triplets_dir[dir])):
            ves = ls_mlai[:,ls_triplets_dir[dir][tri][0], ls_triplets_dir[dir][tri][1], ls_triplets_dir[dir][tri][2]]
            res_trans, res_abs_i = calc_extinc_ray_multi(ves, I0*ls_poids[dir], ls_k[:,dir])#[1.,0.6]) #pour korcer des k 
            
            res_trans_form[ls_triplets_dir[dir][tri][0], ls_triplets_dir[dir][tri][1], ls_triplets_dir[dir][tri][2]] += res_trans
            res_abs_i_form[:, ls_triplets_dir[dir][tri][0], ls_triplets_dir[dir][tri][1], ls_triplets_dir[dir][tri][2]] += res_abs_i
    
    return res_trans_form, res_abs_i_form

#sortir distribution de I0/ls_poids et ls_k (a passer en argument) pour une fonction fortan uniquement calcul


def calc_extinc_allray_multi_reduced(ls_mlai, ls_triplets_dir, ls_distf, I0, optsky=None, opt='VXpXmYpYm'):
    """ """

    # combien/quelles lignes a zeros de LAI au dessus
    laicum = np.sum(ls_mlai, axis=0)
    laicumvert = np.sum(laicum, axis=(1, 2))
    nb0 = 0  # nb de couches sans feuilles/LAI
    for i in range(len(laicumvert)):
        if laicumvert[i] == 0.:
            nb0 += 1
        else:
            break

    # ajouter un if sur nb0 ou le faire a chaque fois?
    # redim des m_lai et triplets
    shp = np.shape(ls_mlai)
    reduced_mlai = []
    for plt in range(shp[0]):
        reduced_mlai.append(ls_mlai[plt, (nb0 - 1):shp[1], :, :])

    reduced_triplets = get_ls_triplets(reduced_mlai[0], opt)
    reduced_mlai = array(reduced_mlai)

    # ls_distf = [riri.disttetaf(abs(45.), 0.), riri.disttetaf(abs(45.), 0.)]
    res_trans_form_red, res_abs_i_form_red = calc_extinc_allray_multi(reduced_mlai, reduced_triplets, ls_distf, I0, optsky)

    # expand matrices de sortie de nb0-1 couches sans faire les calculs
    shpnew = np.shape(res_abs_i_form_red)
    # zz = zeros((shpnew[0], (nb0-1), shpnew[2], shpnew[3]))
    # oo = ones(((nb0-1), shpnew[2], shpnew[3]))*res_trans_form_red[0,0,0]
    res_trans_form = np.concatenate((ones(((nb0 - 1), shpnew[2], shpnew[3])) * res_trans_form_red[0, 0, 0], res_trans_form_red), axis=0)
    res_abs_i_form = np.concatenate((zeros((shpnew[0], (nb0 - 1), shpnew[2], shpnew[3])), res_abs_i_form_red), axis=1)

    return res_trans_form, res_abs_i_form



def k_teta_DC(mean_incl, elevations=[9.23,10.81,26.57,31.08,47.41,52.62,69.16,90]):
    #computes extinction coefficient from mean inclination
    # equations from SIRASCA. Calculations for each elevation angle (p210-211 crop stucture and light microclimate)
    # default elevation for turtle 46 directions 
    #elevations (beam, degre)-> 90 = vertical ; 0=horizontal
    #mean_incl (feuille, degre) -> 90 = vertical ; 0=horizontal
    nh=len(elevations)
    xinc = radians(mean_incl)
    lst_xk=[]
    for ih in elevations:
        hh = radians(ih)
        if (hh >=xinc):
            xk = cos(xinc)
        else:
            xmm = -tan(hh)/tan(xinc)
            #print xmm
            xmm = arccos(xmm)
            xk = cos(xinc)*(2*(xmm - tan(xmm))/pi - 1.)

        lst_xk.append(xk)
    return lst_xk
  

#k_teta_DC(45)
#k_teta_DC(25., elevations=[5., 15., 25., 35., 45., 55., 65., 75., 85.])

def k_teta_distf(teta_beam, dist_class_teta_f):
    """ pour tenir compte d'une distribution d'incli quelconque """
    # teta_beam, ang_class_f (en degre, par rapport au plan horizontal)
    #dist_class_teta_f: relative distribution des surface par classe d'incli, tous les 10 degre, centre [5,15,25,35,45,55,65,75,85]
    ang_class_f = [5,15,25,35,45,55,65,75,85] 
    res = []
    for i in range(len(ang_class_f)):
        kf = k_teta_DC(float(ang_class_f[i]), [teta_beam])[0]*dist_class_teta_f[i]
        res.append(kf)

    return sum(res)

#### R:FR L Faverjon

def schnute(x,a,b,c,d,x1,x2):
    """fonction schnute generale appliquee a x"""
    Y=pow((pow(c,b)+(pow(d,b)-pow(c,b)))*((1-exp(-a*(x-x1)))/(1-exp(-a*((x2-x1))))),(1/b))
    return Y

def rfr_calc_relatif(relI0,a=3.09,b=1.59,c=0,d=1.12,x1=0.,x2=2.):
    """calcul du ratio rouge clair:rouge sombre narrowband a partir du PAR transmis relatif par rapport au rayonnement incident (relI0)
    les parametres par defaut a=3.09,b=1.59,c=0,d=1.12,x1=0.,x2=2. de la fonction schnute sont ajustes sur un couvert de sorgho, 
    Escobar et al 2009 - Agricultural and Forest Meteorology, 149(8), 1244-1253  """
    return schnute(relI0,a,b,c,d,x1,x2)

#applique a la matrice des transmis res_trans_form
#res_rfr = rfr_calc_relatif(res_trans_form)


def disttetaf(mf, sdf, nbs=10000,seed=0):
    """ python version of dist_tetaf.r"""
    #mf = 5.
    #sdf = 10.
    #n = 100#00
    #seed = 0
    
    #random.seed(seed)
    x = normal(loc=mf, scale=sdf, size=nbs) #numpy.random.normal
    x
    x[x<0] = -x[x<0]
    x[x>90] = 90 - (x[x>90] - 90)
    x[x < 0] = -x[x < 0]
    x[x > 90] = 90 - (x[x > 90] - 90)
    x
    res = histogram(x, bins=[0, 10, 20, 30, 40, 50, 60, 70, 80, 90])
    return res[0] / float(nbs)
    #print 't', disttetaf(5., 10.)


